### PR TITLE
TextEdit: Search

### PR DESCRIPTION
**Pull request by JeppeSRC**
Searching in files. 

Features:
- Case sensitivity 
- Match whole word only

The UI is very simple and looks like shit, it's only there as a place holder atm. Someone with a sense for UI can redo it later.

### CODE REVIEW

File: src/Geno/C++/GUI/Widgets/TextEdit.cpp [hunk 1]
```diff
@@ -3439,3 +3507,590 @@ std::vector< int > TextEdit::CursorsNotInText( File& rFile )
...
+					rSearchString[ i ] = '\t';
+					break;
+				default:
+					continue;
+			}
+
+			rSearchString.erase( rSearchString.begin() + i + 1 );
+		}
+	}
+} // PrepareSearchString
```
Comment (Gaztin) [inline src/Geno/C++/GUI/Widgets/TextEdit.cpp:3550]: What if we are searching for a backslash followed by an 'n' character, and not a newline? We can't type double backslashes to escape it. Should be always escape the following character after a backslash maybe?

File: src/Geno/C++/GUI/Widgets/TextEdit.cpp [hunk 2]
```diff
@@ -3439,3 +3507,590 @@ std::vector< int > TextEdit::CursorsNotInText( File& rFile )
...
+char toLower( char c )
+{
+	//TODO: Possibly handle other languages as well
+	if( c >= 'A' && c <= 'Z' )
+	{
+		return c + 32;
+	}
+
+	return c;
+}
```
Comment (Gaztin) [inline src/Geno/C++/GUI/Widgets/TextEdit.cpp:3563]: There's already https://en.cppreference.com/w/cpp/locale/tolower that supports other locales even!

File: src/Geno/C++/GUI/Widgets/TextEdit.h [hunk 3]
```diff
@@ -144,6 +149,76 @@ class TextEdit
...
+
+			std::thread         Thread;
+			std::string         SearchTerm;
+			SearchResultGroups* Result;
+			int                 State;
+		};
+
+		std::vector< SearchInstance* > SearchInstances;
+		SearchResultGroups*            SearchResult = nullptr;
+	};
```
Comment (Gaztin) [inline src/Geno/C++/GUI/Widgets/TextEdit.h:220]: Looks like this should be refactored into its own class

File: src/Geno/C++/GUI/Widgets/TextEdit.cpp (hunk 2 above)
Comment (JeppeSRC) [reply src/Geno/C++/GUI/Widgets/TextEdit.cpp:3563]: That's what I used in the beginning until I realized it was really slow. Like it totally kills performance.

File: src/Geno/C++/GUI/Widgets/TextEdit.h (hunk 3 above)
Comment (JeppeSRC) [reply src/Geno/C++/GUI/Widgets/TextEdit.h:220]: But it's only ever used in textedit. Makes most sense it lives in textedit IMO.

File: src/Geno/C++/GUI/Widgets/TextEdit.cpp (hunk 2 above)
Comment (JeppeSRC) [reply src/Geno/C++/GUI/Widgets/TextEdit.cpp:3563]: Wait! It only seems rubbish when you pass it the locale.

Comment (Gaztin) [reply src/Geno/C++/GUI/Widgets/TextEdit.cpp:3563]: Ah, yeah there seems to be another version that doesn't take a locale

File: src/Geno/C++/GUI/Widgets/TextEdit.h (hunk 3 above)
Comment (Gaztin) [reply src/Geno/C++/GUI/Widgets/TextEdit.h:220]: While I agree, it also takes up a lot of real estate and makes the file harder to skim over.

Comment (JeppeSRC) [reply src/Geno/C++/GUI/Widgets/TextEdit.h:220]: True. So do it get it's own file or just outside the textedit class?

Comment (Gaztin) [reply src/Geno/C++/GUI/Widgets/TextEdit.h]: its own file, preferrably

Comment (JeppeSRC) [reply src/Geno/C++/GUI/Widgets/TextEdit.h]: well that's gonna be a massive pain as it depends on thing in textedit

Comment (Gaztin) [reply src/Geno/C++/GUI/Widgets/TextEdit.h]: All right, keep it this way for now. I'll have to figure out a way for us to split up files in the future.

**Pull request merged**